lib/services/local/presence.go

```diff
@@ -1587,6 +1599,112 @@ func (s *PresenceService) ListResources(ctx context.Context, req proto.ListResou
 	return resources, nextKey, nil
 }
 
+// listResourcesWithSort supports sorting by falling back to retrieving all resources
+// with GetXXXs, filter, and then fake pagination.
+func (s *PresenceService) listResourcesWithSort(ctx context.Context, req proto.ListResourcesRequest) ([]types.ResourceWithLabels, string, error) {
+	var resources []types.ResourceWithLabels
+	switch req.ResourceType {
+	case types.KindNode:
+		nodes, err := s.GetNodes(ctx, req.Namespace)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		servers := types.Servers(nodes)
+		if err := servers.SortByCustom(req.SortBy); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		resources = servers.AsResources()
+
+	case types.KindAppServer:
+		appservers, err := s.GetApplicationServers(ctx, req.Namespace)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		servers := types.AppServers(appservers)
+		if err := servers.SortByCustom(req.SortBy); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		resources = servers.AsResources()
+
+	case types.KindDatabaseServer:
+		dbservers, err := s.GetDatabaseServers(ctx, req.Namespace)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+
+		servers := types.DatabaseServers(dbservers)
+		if err := servers.SortByCustom(req.SortBy); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		resources = servers.AsResources()
+
+	default:
+		return nil, "", trace.NotImplemented("resource type %q is not supported for ListResourcesWithSort", req.ResourceType)
+	}
+
+	return FakePaginate(resources, req)
+}
+
+// FakePaginate is used when we are working with an entire list of resources upfront but still requires pagination.
+func FakePaginate(resources []types.ResourceWithLabels, req proto.ListResourcesRequest) ([]types.ResourceWithLabels, string, error) {
+	limit := int(req.Limit)
+	if limit <= 0 {
+		return nil, "", trace.BadParameter("nonpositive limit value")
+	}
+
+	// Trim resources that precede start key.
+	if req.StartKey != "" {
+		pageStart := 0
+		for i, resource := range resources {
```

Comment:
Can we use `sort.Search` to find the index instead?

Reply:
i looked into it, got initially excited, but i don't think i can... the object field that we search for has to be the same as the fields we sorted by, so for example, if we sort by `db.types` and we need to search for `db.Name`, it won't work.

Reply:
Oh, I see what you mean; the resources are sorted by the `SortBy` but we have to search by pagination key